### PR TITLE
Optimal quasi-probability representations.

### LINKED ISSUE #374: Manually solve for QPR decomposition given noisy gateset
Following the trend of #370, we would like to be able to compute the quasi-probability distribution (QPR) of an "ideal gate" in a more general fashion, without assuming there exists some well-behaved/simple noise model to describe our QPU. In this case, we might be given some description of a "noisy implementable gateset" (e.g. as superoperators or process matrices). Using this description, we can solve the linear problem described in [Temme2017](https://arxiv.org/abs/1612.02058) to find the decomposition with minimum overhead (defined as 1-norm of the QPR coefficients).

### PR DESCRIPTION
**Pull request by andreamari**
Description
-----------
Adds tools for finding the optimal quasi-probability representation of an ideal operation given a noisy basis.
Fixes #374.

Checklist
-----------

Check off the following once complete (or if not applicable) after opening the PR. The PR will be reviewed once this checklist is complete and all tests are passing.

- [x] I added unit tests for new code.
- [x] I used [type hints](https://www.python.org/dev/peps/pep-0484/) in function signatures.
- [x] I used [Google-style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) docstrings for functions.
- [x] I [updated the documentation](../blob/master/docs/CONTRIBUTING_DOCS.md) where relevant.

### CODE REVIEW

**Comment by andreamari** [comment]
sorry @rmlarose, I did a mess with these 3 functions: 
```
    global_depolarizing_kraus,
    local_depolarizing_kraus,
    amplitude_damping_kraus,
```

I this PR, I added them in the corresponding files `representations/depolarizing.py` and  `representations/damping.py` . 

However I realized that they were already in my previous PR (#694) but placed in the file `pec/channels.py`. 
So we now have the same functions in two places :disappointed: .

So, instead of creating additional confusion with new changes, I would ask you do suggest what is an optimal location for the functions above (`pec/channels.py` or `pec/representations/depolarizing(damping).py`). After this I will remove one of the two copies and update test files consistently.

File: mitiq/pec/representations/damping.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
+# Copyright (C) 2020 Unitary Fund
```
Comment (rmlarose) [inline mitiq/pec/representations/damping.py:1]:
```suggestion
# Copyright (C) 2021 Unitary Fund
```

File: mitiq/pec/representations/damping.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    Any ideal single-qubit unitary followed by local amplitude-damping noise
+    of equal ``noise_level`` is assumed to be in the basis of implementable
+    operations.
+
+    The representation is based on the analytical result presented
+    in :cite:`Takagi2020`.
+
+    Args:
+        ideal_operation: The ideal operation (as a QPROGRAM) to represent.
+        noise_level: The noise level of each depolarizing channel.
```
Comment (rmlarose) [inline mitiq/pec/representations/damping.py:52]:
```suggestion
        noise_level: The noise level of each amplitude damping channel.
```

File: mitiq/pec/representations/damping.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    # Build basis_expantion
+    expansion = {NoisyOperation(c): a for c, a in zip(imp_op_circuits, etas)}
+
+    return OperationRepresentation(ideal_operation, expansion)
+
+
+def amplitude_damping_kraus(
+    noise_level: float, num_qubits: int,
+) -> List[np.ndarray]:
+    """Returns the kraus operators of the tensor product of local
```
Comment (rmlarose) [inline mitiq/pec/representations/damping.py:101]:
```suggestion
    """Returns the Kraus operators of the tensor product of local
```

File: mitiq/pec/representations/optimal.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+"""Functions for finding optimal representations given a noisy basis."""
+
+from typing import List
+from mitiq import QPROGRAM
+from mitiq.pec.types import NoisyBasis, OperationRepresentation
+import numpy as np
+from scipy.optimize import minimize, LinearConstraint
+from mitiq.pec.channels import matrix_to_vector, kraus_to_super
+from cirq import channel
+from mitiq.conversions import convert_to_mitiq
```
Comment (rmlarose) [inline mitiq/pec/representations/optimal.py:25]:
```suggestion
from typing import List

import numpy as np
from scipy.optimize import minimize, LinearConstraint

from cirq import channel

from mitiq import QPROGRAM
from mitiq.conversions import convert_to_mitiq
from mitiq.pec.types import NoisyBasis, OperationRepresentation
from mitiq.pec.channels import matrix_to_vector, kraus_to_super
```

Comment (rmlarose) [reply mitiq/pec/representations/optimal.py:25]: (to be [PEP compliant](https://www.python.org/dev/peps/pep-0008/#imports))

File: mitiq/pec/representations/optimal.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        [matrix_to_vector(mat) for mat in basis_matrices_real]
+    ).T
+    array_b = matrix_to_vector(ideal_matrix_real)
+
+    constraint = LinearConstraint(matrix_a, lb=array_b - tol, ub=array_b + tol)
+
+    def one_norm(x):
+        return np.linalg.norm(x, 1)
+
+    initial_guess = np.zeros(len(basis_matrices))
```
Comment (rmlarose) [inline mitiq/pec/representations/optimal.py:78]: Allow for optional initial guess as input argument?

File: mitiq/pec/representations/optimal.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
+# Copyright (C) 2020 Unitary Fund
```
Comment (rmlarose) [inline mitiq/pec/representations/optimal.py:1]:
```suggestion
# Copyright (C) 2021 Unitary Fund
```

File: mitiq/pec/representations/tests/test_damping.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
+# Copyright (C) 2020 Unitary Fund
```
Comment (rmlarose) [inline mitiq/pec/representations/tests/test_damping.py:1]:
```suggestion
# Copyright (C) 2021 Unitary Fund
```

File: mitiq/pec/representations/tests/test_optimal.py [hunk 8]
```diff
@@ -0,0 +1,361 @@
+# Copyright (C) 2020 Unitary Fund
```
Comment (rmlarose) [inline mitiq/pec/representations/tests/test_optimal.py:1]:
```suggestion
# Copyright (C) 2021 Unitary Fund
```

File: mitiq/pec/representations/tests/test_optimal.py [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+    choi_to_super,
+)
+
+from mitiq.pec.types import OperationRepresentation, NoisyBasis, NoisyOperation
+from mitiq.conversions import convert_from_mitiq
+
+from mitiq.utils import _equal
+
+
+def _equivalent_representations(
```
Comment (rmlarose) [inline mitiq/pec/representations/tests/test_optimal.py:67]: I think we should move this to `OperationRepresentation.__eq__`. Thoughts?

File: mitiq/pec/representations/tests/test_optimal.py [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+        # and should also increase the objective function
+        assert worst_case_error < previous_error
+        assert minimum > previous_minimum
+        previous_error = worst_case_error
+        previous_minimum = minimum
+
+
+@mark.parametrize("circ_type", ["cirq", "qiskit", "pyquil", "braket"])
+def test_find_optimal_representation_depolarizing_two_qubit_gates(circ_type):
+    """Test optimal representation of agrees with a known analytic result."""
```
Comment (rmlarose) [inline mitiq/pec/representations/tests/test_optimal.py:221]:
```suggestion
    """Test optimal representation agrees with a known analytic result."""
```

File: mitiq/pec/representations/tests/test_optimal.py [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+        expected_rep = represent_operation_with_local_depolarizing_noise(
+            ideal_op_native, noise_level,
+        )
+        assert np.allclose(np.sort(rep.coeffs), np.sort(expected_rep.coeffs))
+        assert _equivalent_representations(rep, expected_rep)
+
+
+@mark.parametrize("circ_type", ["cirq", "qiskit", "pyquil", "braket"])
+def test_find_optimal_representation_single_qubit_depolarizing(circ_type):
+    """Test optimal representation of agrees with a known analytic result."""
```
Comment (rmlarose) [inline mitiq/pec/representations/tests/test_optimal.py:269]:
```suggestion
    """Test optimal representation agrees with a known analytic result."""
```

**Review by rmlarose — APPROVED** [review]
A few minor comments but LGTM @andreamari. Awesome stuff.

File: mitiq/pec/representations/optimal.py (hunk 5 above)
Comment (andreamari) [reply mitiq/pec/representations/optimal.py]: Added here: 5a649f6

File: mitiq/pec/representations/tests/test_optimal.py (hunk 9 above)
Comment (andreamari) [reply mitiq/pec/representations/tests/test_optimal.py]: Makes sense. Added here bfab2cd

**Comment by rmlarose** [comment]
> sorry @rmlarose, I did a mess with these 3 functions:
> 
> ```
>     global_depolarizing_kraus,
>     local_depolarizing_kraus,
>     amplitude_damping_kraus,
> ```
> 
> I this PR, I added them in the corresponding files `representations/depolarizing.py` and `representations/damping.py` .
> 
> However I realized that they were already in my previous PR (#694) but placed in the file `pec/channels.py`.
> So we now have the same functions in two places  .
> 
> So, instead of creating additional confusion with new changes, I would ask you do suggest what is an optimal location for the functions above (`pec/channels.py` or `pec/representations/depolarizing(damping).py`). After this I will remove one of the two copies and update test files consistently.

Oof, how did I miss this. Let's remove them from this PR, then LGTM to merge. Thanks for addressing my comments.

**Review by rmlarose — APPROVED** [review]
Actually, I see you removed the duplicate functions from pec.channels, correct? If so, LGTM.

**Pull request merged**